メモの保存先をローカルストレージからDBに変更

IkumaTadokoro opened the pull request:
## 概要

- FJORD BOOT CAMP 「WebアプリからのDB利用」課題提出物
- メモの保存先をローカルストレージ（JSONファイル）から、PostgreSQL（`memo`テーブル）に変更

## 技術的変更点概要

- 以下の各処理について、ローカルストレージではなくDBへアクセスするように変更                    
  - 一覧表示（`index`）
  - 詳細表示(`show`)
  - 新規登録(`create`)
  - 更新(`update`)
  - 削除(`destroy`)

## 使い方

- 環境設定
  - PostgreSQLへの接続情報（環境変数）は、`./env`に以下の通り設定する。

```
PG_HOST='ホスト名'
PG_PORT='ポート番号'
PG_USER='ユーザ名'
PGPASSWORD='パスワード'
PGDATABASE='データベース名'
```

- インストール

```bash
$ bundle exec app.rb
```

## UIに対する変更

- なし

## マイグレーション

- `memo`テーブル

```sql

create table memo
(
    id      serial not null,
    title   varchar(100),
    content varchar(200)
);

```

IkumaTadokoro committed:
Add gem 'pg'

IkumaTadokoro committed:
Add settings for 'pg'

IkumaTadokoro committed:
Use database as a storage instead of local json file in index

IkumaTadokoro committed:
Use database as a storage instead of local json file in show

IkumaTadokoro committed:
Use database as a storage instead of local json file in create

IkumaTadokoro committed:
Use database as a storage instead of local json file in update

IkumaTadokoro committed:
Use database as a storage instead of local json file in destroy

IkumaTadokoro committed:
Remove unused require and constant

IkumaTadokoro committed:
Remove "ORDER BY updated_at" not mentioned in the specifications

IkumaTadokoro committed:
Add gem 'dotenv'

IkumaTadokoro committed:
Add .gitignore

IkumaTadokoro committed:
Use environment variables as connection string

IkumaTadokoro committed:
Remove unused converting hash key into symbol at Memo.index

IkumaTadokoro committed:
Change local variable name in Memo.show

IkumaTadokoro committed:
Remove unused converting key into symbol at Memo.show

IkumaTadokoro committed:
Use instance variable in Memo#destroy

IkumaTadokoro committed:
Use instance variable in Memo#update

betachelsea commented on lib/memo.rb:
```diff
@@ -15,30 +22,24 @@ def initialize(id:, title:, content:)
 
   def destroy
-    File.delete("#{MEMO_DIR}/#{@id}.json")
+    @@connection.exec('DELETE FROM memo WHERE id = $1', [@id])
```
SQLインジェクション対策がされていてgoodです！ 👍

betachelsea commented on lib/memo.rb:
```diff
@@ -1,10 +1,17 @@
+      port: ENV['PG_PORT'],
+      user: ENV['PG_USER'],
+      password: ENV['PGPASSWORD'],
+      dbname: ENV['PGDATABASE']
```
環境変数へ認証情報が逃がされていてここもナイスですね 👍

betachelsea commented on lib/memo.rb:
```diff
@@ -1,10 +1,17 @@
 
 class Memo
-  MEMO_DIR = './memos'
+  @@connection = PG.connect(
```
DBとのコネクションは「開けたら閉める」が望ましいです！ 👀 

今回の実装だと、一度開けたらアプリが終了するまで開けっぱなしになってしまっているようなので、各動作時に  開ける→(処理する)→ 閉める を実施できるように検討してみてほしいです 🙏

IkumaTadokoro replied:
> 各動作時に 開ける→(処理する)→ 閉める を実施できるように検討してみてほしいです 🙏

`exec_sql`メソッド（open→ブロック実行→closeを行う）を定義し、各SQL実行ごとにコネクションをクローズするように修正いたしました。ご確認よろしくお願いいたします🙇‍♂️

betachelsea reviewed (commented):
@IkumaTadokoro レビューしました！

IkumaTadokoro committed:
Define `exec_sql` for managing DB connection

IkumaTadokoro committed:
Use PG::Connection#exec_params instead of PG::Connection#exec

```
For backward compatibility, if you pass more than one parameter to this method, it will call #exec_params for you. New code should explicitly use #exec_params if argument placeholders are used.
```

betachelsea reviewed (commented):
修正ありがとうございます！LGTMです 👍
